# AI Financial Operations & Customer Service Platform

## V0.1 - Day 1: Data Processing & Data Quality

### Objective

The objective of Day 1 is to prepare the financial datasets for further
analysis and AI processing.

The following activities will be performed:

1. Load the datasets
2. Understand the dataset structure
3. Check data types
4. Identify missing values
5. Identify duplicate records
6. Identify invalid values
7. Validate relationships between datasets
8. Clean the data
9. Re-check data quality
10. Save the cleaned datasets

In [1]:
import pandas as pd
import numpy as np

from pathlib import Path

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

## 2. Define Dataset Paths

The project contains four relational datasets:

- Customers
- Transactions
- Service Cases
- Fraud Alerts

In [2]:
data_path = Path(".")
customers_file = data_path / "customers.csv"
transactions_file = data_path / "transactions.csv"
service_cases_file = data_path / "service_cases.csv"
fraud_alerts_file = data_path / "fraud_alerts.csv"

## 3. Load Datasets

The four CSV files are loaded into Pandas DataFrames for processing.

In [3]:
customers = pd.read_csv(customers_file)
transactions = pd.read_csv(transactions_file)
service_cases = pd.read_csv(service_cases_file)
fraud_alerts = pd.read_csv(fraud_alerts_file)

In [4]:
print("Customers:", customers.shape)
print("Transactions:", transactions.shape)
print("Service Cases:", service_cases.shape)
print("Fraud Alerts:", fraud_alerts.shape)

Customers: (1200, 9)
Transactions: (7015, 10)
Service Cases: (2500, 11)
Fraud Alerts: (1100, 8)


## 4. Dataset Overview

This section examines the structure, columns, data types and sample
records of each dataset.

In [5]:
customers.head()

,customer_id,full_name,age,gender,city,account_type,customer_segment,kyc_status,onboarding_date
0,CUST00001,Customer 1,18.0,Male,Bengaluru,Current,Retail,Verified,2025-12-09
1,CUST00002,Customer 2,57.0,Female,Mumbai,Savings,Retail,Verified,2023-11-19
2,CUST00003,Customer 3,52.0,Female,Pune,Savings,Retail,Verified,2026-01-10
3,CUST00004,Customer 4,21.0,Male,Kolkata,Savings,Retail,Verified,2024-07-28
4,CUST00005,Customer 5,70.0,Female,Hyderabad,Savings,Retail,Verified,2022-01-01


In [6]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customer_id       1200 non-null   object 
 1   full_name         1200 non-null   object 
 2   age               1186 non-null   float64
 3   gender            1200 non-null   object 
 4   city              1179 non-null   object 
 5   account_type      1200 non-null   object 
 6   customer_segment  1200 non-null   object 
 7   kyc_status        1182 non-null   object 
 8   onboarding_date   1200 non-null   object 
dtypes: float64(1), object(8)
memory usage: 84.5+ KB


In [7]:
customers.describe(include="all")

,customer_id,full_name,age,gender,city,account_type,customer_segment,kyc_status,onboarding_date
count,1200,1200,1186.000000,1200,1179,1200,1200,1182,1200
unique,1200,1200,NaN,3,15,2,4,3,901
top,CUST00001,Customer 1,NaN,Male,Vijayawada,Savings,Retail,Verified,2022-06-30
freq,1,1,NaN,611,148,1013,752,1005,4
mean,NaN,NaN,46.158516,NaN,NaN,NaN,NaN,NaN,NaN
std,NaN,NaN,16.780142,NaN,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,18.000000,NaN,NaN,NaN,NaN,NaN,NaN
25%,NaN,NaN,32.000000,NaN,NaN,NaN,NaN,NaN,NaN
50%,NaN,NaN,45.000000,NaN,NaN,NaN,NaN,NaN,NaN
75%,NaN,NaN,61.000000,NaN,NaN,NaN,NaN,NaN,NaN


## 6. Missing Value Analysis

The Customers dataset contains missing values in three columns:
age, city, and kyc_status.

Missing values will be analyzed before deciding the appropriate
data-cleaning strategy.

In [8]:
customers.isnull().sum()

customer_id          0
full_name            0
age                 14
gender               0
city                21
account_type         0
customer_segment     0
kyc_status          18
onboarding_date      0
dtype: int64

In [9]:
missing_customers = pd.DataFrame({
    "missing_count": customers.isnull().sum(),
    "missing_percentage": customers.isnull().mean() * 100
})

missing_customers

,missing_count,missing_percentage
customer_id,0,0.000000
full_name,0,0.000000
age,14,1.166667
gender,0,0.000000
city,21,1.750000
account_type,0,0.000000
customer_segment,0,0.000000
kyc_status,18,1.500000
onboarding_date,0,0.000000


In [10]:
print("Duplicate rows:", customers.duplicated().sum())
print("Duplicate customer IDs:", customers["customer_id"].duplicated().sum())

Duplicate rows: 0
Duplicate customer IDs: 0


## 8. Transactions Dataset - Data Quality Analysis

The transactions dataset is analyzed to identify its structure,
missing values, duplicate records, duplicate transaction IDs,
and potential data-quality issues.

In [11]:
transactions.shape

(7015, 10)

In [12]:
transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7015 entries, 0 to 7014
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   transaction_id         7015 non-null   object 
 1   customer_id            6973 non-null   object 
 2   transaction_timestamp  7015 non-null   object 
 3   transaction_type       7015 non-null   object 
 4   amount                 6959 non-null   float64
 5   merchant_category      6945 non-null   object 
 6   payment_channel        6931 non-null   object 
 7   transaction_status     6945 non-null   object 
 8   is_international       7015 non-null   int64  
 9   failed_attempts        7015 non-null   int64  
dtypes: float64(1), int64(2), object(7)
memory usage: 548.2+ KB


In [13]:
transactions.head()

,transaction_id,customer_id,transaction_timestamp,transaction_type,amount,merchant_category,payment_channel,transaction_status,is_international,failed_attempts
0,TXN0000001,CUST00682,2025-07-23 22:00:00,Credit,1779.86,Entertainment,Net Banking,Failed,0,0
1,TXN0000002,CUST01010,2025-01-26 22:00:00,Transfer,982.17,Entertainment,Credit Card,Success,0,0
2,TXN0000003,CUST01002,2025-02-24 22:00:00,Debit,5741.03,Shopping,Debit Card,Success,0,0
3,TXN0000004,CUST00654,2026-02-17 14:00:00,Debit,3797.44,Food,Debit Card,Failed,0,0
4,TXN0000005,CUST00015,2026-01-24 07:00:00,Debit,873.60,Rent,ATM,Success,0,1


In [14]:
transactions.isnull().sum()

transaction_id            0
customer_id              42
transaction_timestamp     0
transaction_type          0
amount                   56
merchant_category        70
payment_channel          84
transaction_status       70
is_international          0
failed_attempts           0
dtype: int64

In [15]:
missing_transactions = pd.DataFrame({
    "missing_count": transactions.isnull().sum(),
    "missing_percentage": transactions.isnull().mean() * 100
})

missing_transactions

,missing_count,missing_percentage
transaction_id,0,0.000000
customer_id,42,0.598717
transaction_timestamp,0,0.000000
transaction_type,0,0.000000
amount,56,0.798289
merchant_category,70,0.997862
payment_channel,84,1.197434
transaction_status,70,0.997862
is_international,0,0.000000
failed_attempts,0,0.000000


In [16]:
print(
    "Duplicate rows:",
    transactions.duplicated().sum()
)

Duplicate rows: 15


In [17]:
print(
    "Duplicate transaction IDs:",
    transactions["transaction_id"].duplicated().sum()
)

Duplicate transaction IDs: 15


In [18]:
transactions.describe()

,amount,is_international,failed_attempts
count,6959.000000,7015.000000,7015.000000
mean,1899.533413,0.058019,0.250178
std,2629.263134,0.233795,0.497490
min,-500.000000,0.000000,0.000000
25%,534.265000,0.000000,0.000000
50%,1096.850000,0.000000,0.000000
75%,2234.690000,0.000000,0.000000
max,43312.140000,1.000000,3.000000


## 8.1 Transaction Categorical Value Validation

The categorical fields are checked for unique values and unexpected
categories before applying any standardization or cleaning.

In [19]:
transactions["transaction_type"].value_counts(dropna=False)

transaction_type
Debit       3498
Transfer    2101
Credit      1416
Name: count, dtype: int64

In [20]:
transactions["merchant_category"].value_counts(dropna=False)

merchant_category
Travel           741
Entertainment    731
Salary           716
Groceries        714
Shopping         712
Investment       708
Rent             672
Food             669
Utilities        645
Healthcare       637
NaN               70
Name: count, dtype: int64

In [21]:
transactions["payment_channel"].value_counts(dropna=False)

payment_channel
UPI            2975
Debit Card     1388
Net Banking     953
Credit Card     880
ATM             719
NaN              84
upi              12
credit_card       2
debit_card        2
Name: count, dtype: int64

In [22]:
transactions["transaction_status"].value_counts(dropna=False)

transaction_status
Success    6032
Failed      561
Pending     352
NaN          70
Name: count, dtype: int64

In [23]:
transactions["is_international"].value_counts(dropna=False)

is_international
0    6608
1     407
Name: count, dtype: int64

## 8.2 Negative Transaction Amount Validation

Negative transaction amounts are identified for further investigation.
They are not removed automatically because the business meaning of a
negative transaction must first be confirmed.

In [24]:
negative_transactions = transactions[
    transactions["amount"] < 0
]

print("Negative transaction records:", len(negative_transactions))

negative_transactions[
    [
        "transaction_id",
        "customer_id",
        "transaction_type",
        "amount",
        "merchant_category",
        "transaction_status"
    ]
].head(20)

Negative transaction records: 12


,transaction_id,customer_id,transaction_type,amount,merchant_category,transaction_status
212,TXN0000213,CUST01043,Debit,-1.0,Investment,Success
320,TXN0000321,CUST00854,Debit,-100.0,Food,Success
803,TXN0000804,CUST01122,Debit,-500.0,Travel,Success
2259,TXN0002260,CUST00586,Debit,-500.0,Entertainment,Success
3536,TXN0003537,CUST01053,Credit,-1.0,Travel,Success
3588,TXN0003589,CUST01200,Debit,-1.0,Entertainment,Success
4031,TXN0004032,CUST01124,Debit,-1.0,Shopping,Success
5835,TXN0005836,CUST01118,Credit,-100.0,Healthcare,Success
5954,TXN0005955,CUST01024,Credit,-500.0,Healthcare,Success
6406,TXN0006407,CUST00047,Credit,-1.0,Healthcare,Success


In [25]:
zero_transactions = transactions[
    transactions["amount"] == 0
]

print("Zero amount transactions:", len(zero_transactions))

Zero amount transactions: 6


In [26]:
transactions[
    transactions["customer_id"].isna()
].head(20)

,transaction_id,customer_id,transaction_timestamp,transaction_type,amount,merchant_category,payment_channel,transaction_status,is_international,failed_attempts
324,TXN0000325,NaN,2025-02-21 13:00:00,Debit,1336.28,Healthcare,ATM,Failed,0,1
510,TXN0000511,NaN,2025-02-08 01:00:00,Debit,3401.51,Utilities,UPI,Success,0,0
790,TXN0000791,NaN,2026-09-25 07:00:00,Credit,307.88,Rent,UPI,Success,0,0
814,TXN0000815,NaN,2026-03-20 18:00:00,Transfer,2041.28,Food,Debit Card,Pending,0,0
889,TXN0000890,NaN,2025-04-04 21:00:00,Credit,598.50,Shopping,Net Banking,Success,0,0
1220,TXN0001221,NaN,2026-06-25 12:00:00,Debit,3066.04,Entertainment,Net Banking,Success,0,1
1673,TXN0001674,NaN,2025-03-28 15:00:00,Transfer,1383.08,Food,Debit Card,Success,0,0
1707,TXN0001708,NaN,2025-03-14 21:00:00,Debit,11697.91,Travel,Net Banking,Success,0,0
1826,TXN0001827,NaN,2026-02-21 23:00:00,Credit,1295.25,Groceries,UPI,Pending,0,0
1833,TXN0001834,NaN,2025-06-16 02:00:00,Debit,309.94,Utilities,Debit Card,Success,0,0


### 8.4 Customer ID Referential Integrity

The transaction dataset is expected to reference valid customer records
through the customer_id field.

This check identifies transactions whose customer_id does not exist
in the customers dataset.

In [27]:
invalid_customer_ids = transactions[
    transactions["customer_id"].notna()
    & ~transactions["customer_id"].isin(
        customers["customer_id"]
    )
]

print(
    "Transactions with non-existing customer IDs:",
    len(invalid_customer_ids)
)

Transactions with non-existing customer IDs: 0


In [28]:
invalid_customer_ids[
    [
        "transaction_id",
        "customer_id",
        "amount",
        "transaction_type",
        "transaction_status"
    ]
].head(20)

,transaction_id,customer_id,amount,transaction_type,transaction_status


## 9. Service Cases Dataset - Data Quality Analysis

The service cases dataset is analyzed to identify its structure,
missing values, duplicate records, invalid values, and relationships
with customers and transactions.

In [29]:
service_cases.shape

(2500, 11)

In [30]:
service_cases.info()
service_cases.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2500 entries, 0 to 2499
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   case_id                2500 non-null   object 
 1   customer_id            2500 non-null   object 
 2   transaction_id         2500 non-null   object 
 3   created_at             2500 non-null   object 
 4   case_type              2500 non-null   object 
 5   priority               2463 non-null   object 
 6   status                 2500 non-null   object 
 7   resolution_time_hours  2450 non-null   float64
 8   sla_breached           2500 non-null   bool   
 9   channel                2500 non-null   object 
 10  customer_query         2470 non-null   object 
dtypes: bool(1), float64(1), object(9)
memory usage: 197.9+ KB


,case_id,customer_id,transaction_id,created_at,case_type,priority,status,resolution_time_hours,sla_breached,channel,customer_query
0,CASE000001,CUST00226,TXN0002346,2026-06-21,Card Issue,Low,In Progress,14.08,False,App,My card transaction is not working.
1,CASE000002,CUST00083,TXN0002212,2026-04-28,Transaction Dispute,Medium,Resolved,7.86,False,Chat,I do not recognize this transaction.
2,CASE000003,CUST00214,TXN0006604,2025-10-20,Card Issue,High,Resolved,18.40,False,Email,My card transaction is not working.
3,CASE000004,CUST00577,TXN0003732,2026-01-19,General Enquiry,NaN,Resolved,28.13,True,Email,I need information about my account.
4,CASE000005,CUST00913,TXN0000554,2025-01-15,Refund,High,Resolved,21.36,False,App,When will I receive my refund?


In [31]:
service_cases.isnull().sum()

case_id                   0
customer_id               0
transaction_id            0
created_at                0
case_type                 0
priority                 37
status                    0
resolution_time_hours    50
sla_breached              0
channel                   0
customer_query           30
dtype: int64

In [32]:
missing_service_cases = pd.DataFrame({
    "missing_count": service_cases.isnull().sum(),
    "missing_percentage": service_cases.isnull().mean() * 100
})

missing_service_cases

,missing_count,missing_percentage
case_id,0,0.00
customer_id,0,0.00
transaction_id,0,0.00
created_at,0,0.00
case_type,0,0.00
priority,37,1.48
status,0,0.00
resolution_time_hours,50,2.00
sla_breached,0,0.00
channel,0,0.00


In [33]:
print(
    "Duplicate rows:",
    service_cases.duplicated().sum()
)

Duplicate rows: 0


In [34]:
print(
    "Duplicate case IDs:",
    service_cases["case_id"].duplicated().sum()
)

Duplicate case IDs: 0


In [35]:
service_cases.describe()

,resolution_time_hours
count,2450.000000
mean,18.434984
std,12.172637
min,-9.000000
25%,9.392500
50%,16.230000
75%,24.677500
max,103.910000


### 9.1 Resolution Time Validation

Resolution time is checked for negative values because a service case
cannot normally have a negative resolution duration.

Negative values are identified for investigation before cleaning.

In [36]:
negative_resolution = service_cases[
    service_cases["resolution_time_hours"] < 0
]

print(
    "Negative resolution time records:",
    len(negative_resolution)
)

Negative resolution time records: 12


In [37]:
negative_resolution[
    [
        "case_id",
        "customer_id",
        "transaction_id",
        "case_type",
        "priority",
        "status",
        "resolution_time_hours",
        "sla_breached"
    ]
]

,case_id,customer_id,transaction_id,case_type,priority,status,resolution_time_hours,sla_breached
544,CASE000545,CUST01183,TXN0004889,Card Issue,Low,In Progress,-5.0,False
639,CASE000640,CUST00999,TXN0003326,Account Issue,Low,In Progress,-8.0,True
1099,CASE001100,CUST00437,TXN0005941,Card Issue,Low,Escalated,-6.0,False
1159,CASE001160,CUST00903,TXN0003704,Account Issue,Medium,Resolved,-5.0,False
1161,CASE001162,CUST00731,TXN0000917,Transaction Dispute,Low,Open,-9.0,False
1729,CASE001730,CUST00928,TXN0003423,General Enquiry,Low,In Progress,-8.0,False
1752,CASE001753,CUST00244,TXN0001965,General Enquiry,Medium,Resolved,-6.0,False
1792,CASE001793,CUST00119,TXN0003387,KYC,Medium,Open,-5.0,False
1823,CASE001824,CUST00472,TXN0001620,Card Issue,Medium,In Progress,-8.0,True
2100,CASE002101,CUST00237,TXN0000957,Payment Failure,High,Open,-1.0,True


### 9.2 Service Case Categorical Validation

The categorical fields are checked for their unique values and
frequency before applying any cleaning or standardization.

In [38]:
service_cases["case_type"].value_counts(dropna=False)

case_type
KYC                    396
Card Issue             366
Refund                 355
General Enquiry        353
Account Issue          350
Transaction Dispute    341
Payment Failure        339
Name: count, dtype: int64

In [39]:
service_cases["priority"].value_counts(dropna=False)

priority
Medium      1216
Low          591
High         546
Critical     110
NaN           37
Name: count, dtype: int64

In [40]:
service_cases["status"].value_counts(dropna=False)

status
Resolved       1286
In Progress     611
Open            418
Escalated       180
resolved          5
Name: count, dtype: int64

In [41]:
service_cases["sla_breached"].value_counts(dropna=False)

sla_breached
False    1835
True      665
Name: count, dtype: int64

In [42]:
service_cases["channel"].value_counts(dropna=False)

channel
Chat      549
App       511
Email     505
Branch    471
Phone     464
Name: count, dtype: int64

In [43]:
invalid_case_customers = service_cases[
    ~service_cases["customer_id"].isin(
        customers["customer_id"]
    )
]

print(
    "Service cases with non-existing customer IDs:",
    len(invalid_case_customers)
)

Service cases with non-existing customer IDs: 0


In [44]:
invalid_case_transactions = service_cases[
    ~service_cases["transaction_id"].isin(
        transactions["transaction_id"]
    )
]

print(
    "Service cases with non-existing transaction IDs:",
    len(invalid_case_transactions)
)

Service cases with non-existing transaction IDs: 0


### Service Cases Data Quality Findings

- Dataset contains 2,500 service cases and 11 columns.
- 117 missing values were identified.
- Missing values occur in priority, resolution_time_hours and customer_query.
- No duplicate rows were detected.
- No duplicate case IDs were detected.
- 12 records contain negative resolution times.
- The status column contains an inconsistent lowercase "resolved" value.
- 665 cases have SLA breaches.
- No invalid customer IDs were detected.
- No invalid transaction IDs were detected.

The identified issues will be addressed during the data-cleaning stage.

## 10. Fraud Alerts Dataset - Data Quality Analysis

The fraud alerts dataset is analyzed to identify its structure,
missing values, duplicate records, invalid values, and its relationship
with the transactions dataset.

In [45]:
fraud_alerts.info()
fraud_alerts.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1100 entries, 0 to 1099
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   alert_id              1100 non-null   object 
 1   transaction_id        1100 non-null   object 
 2   alert_timestamp       1100 non-null   object 
 3   anomaly_score         1084 non-null   float64
 4   alert_type            1100 non-null   object 
 5   risk_level            1100 non-null   object 
 6   investigation_status  1087 non-null   object 
 7   analyst_action        1100 non-null   object 
dtypes: float64(1), object(7)
memory usage: 68.9+ KB


,alert_id,transaction_id,alert_timestamp,anomaly_score,alert_type,risk_level,investigation_status,analyst_action
0,ALRT000001,TXN0004029,2025-02-24,0.194,Multiple Failures,Low,New,Escalate
1,ALRT000002,TXN0006211,2026-01-16,0.084,Unusual Location,Low,Under Review,Escalate
2,ALRT000003,TXN0005346,2026-06-27,0.862,Unusual Location,Critical,Under Review,Contact Customer
3,ALRT000004,TXN0005366,2026-06-14,0.248,Unusual Location,Low,Under Review,No Action
4,ALRT000005,TXN0002162,2025-11-06,0.164,Unusual Location,Low,Cleared,No Action


In [46]:
fraud_alerts.shape

(1100, 8)

In [47]:
fraud_alerts.isnull().sum()

alert_id                 0
transaction_id           0
alert_timestamp          0
anomaly_score           16
alert_type               0
risk_level               0
investigation_status    13
analyst_action           0
dtype: int64

In [48]:
missing_fraud_alerts = pd.DataFrame({
    "missing_count": fraud_alerts.isnull().sum(),
    "missing_percentage": fraud_alerts.isnull().mean() * 100
})

missing_fraud_alerts

,missing_count,missing_percentage
alert_id,0,0.000000
transaction_id,0,0.000000
alert_timestamp,0,0.000000
anomaly_score,16,1.454545
alert_type,0,0.000000
risk_level,0,0.000000
investigation_status,13,1.181818
analyst_action,0,0.000000


In [49]:
print(
    "Duplicate rows:",
    fraud_alerts.duplicated().sum()
)

Duplicate rows: 0


In [50]:
print(
    "Duplicate alert IDs:",
    fraud_alerts["alert_id"].duplicated().sum()
)

Duplicate alert IDs: 0


In [51]:
fraud_alerts.describe()

,anomaly_score
count,1084.000000
mean,0.417937
std,0.272018
min,0.008000
25%,0.194000
50%,0.345000
75%,0.658250
max,0.990000


### 10.1 Anomaly Score Validation

The anomaly_score field is expected to represent an anomaly risk score.
Its range is checked to identify values outside the expected 0 to 1 range.

Missing scores are also retained for separate treatment during the
cleaning stage.

In [52]:
invalid_anomaly_scores = fraud_alerts[
    fraud_alerts["anomaly_score"].notna()
    & (
        (fraud_alerts["anomaly_score"] < 0)
        | (fraud_alerts["anomaly_score"] > 1)
    )
]

print(
    "Invalid anomaly scores:",
    len(invalid_anomaly_scores)
)

Invalid anomaly scores: 0


### 10.2 Fraud Alert Categorical Validation

Categorical fields are checked for their unique values and frequencies
to identify missing or inconsistent categories.

In [53]:
fraud_alerts["alert_type"].value_counts(dropna=False)

alert_type
Rapid Transactions        232
Unusual Location          230
International Activity    214
Multiple Failures         212
High Amount               212
Name: count, dtype: int64

In [54]:
fraud_alerts["risk_level"].value_counts(dropna=False)

risk_level
Low         469
Medium      366
High        213
Critical     52
Name: count, dtype: int64

In [55]:
fraud_alerts["investigation_status"].value_counts(dropna=False)

investigation_status
Cleared         401
Under Review    341
New             266
Confirmed        79
NaN              13
Name: count, dtype: int64

In [56]:
fraud_alerts["analyst_action"].value_counts(dropna=False)

analyst_action
No Action            296
Escalate             294
Block Transaction    268
Contact Customer     242
Name: count, dtype: int64

### 10.3 Transaction ID Referential Integrity

Each fraud alert should reference a valid transaction.
This check identifies fraud alerts whose transaction_id does not
exist in the transactions dataset.

In [57]:
invalid_alert_transactions = fraud_alerts[
    ~fraud_alerts["transaction_id"].isin(
        transactions["transaction_id"]
    )
]

print(
    "Fraud alerts with non-existing transaction IDs:",
    len(invalid_alert_transactions)
)

Fraud alerts with non-existing transaction IDs: 8


In [58]:
invalid_alert_transactions[
    [
        "alert_id",
        "transaction_id",
        "anomaly_score",
        "risk_level",
        "investigation_status"
    ]
].head(20)

,alert_id,transaction_id,anomaly_score,risk_level,investigation_status
238,ALRT000239,TXN99995,0.218,Low,Cleared
239,ALRT000240,TXN99990,0.248,Medium,Cleared
258,ALRT000259,TXN99997,0.979,High,Under Review
287,ALRT000288,TXN99993,0.260,Low,Cleared
391,ALRT000392,TXN99994,0.667,High,Under Review
1000,ALRT001001,TXN99996,0.276,Low,Cleared
1001,ALRT001002,TXN99991,0.272,Medium,Under Review
1006,ALRT001007,TXN99992,0.122,Low,Cleared


# 11. Overall Data Quality Summary

The four project datasets were analyzed for missing values,
duplicates, invalid numerical values, categorical inconsistencies,
and referential integrity issues.

The identified issues will be addressed during the data-cleaning
stage. No records are removed during this discovery phase.

| Dataset | Rows | Main Issues Identified |
|---|---:|---|
| Customers | 1,200 | Missing age, city, kyc_status |
| Transactions | 7,015 | Missing values, 15 duplicates, 12 negative amounts, 6 zero amounts, 42 missing customer IDs |
| Service Cases | 2,500 | Missing priority/resolution/query, 12 negative resolution times, 5 inconsistent status values |
| Fraud Alerts | 1,100 | Missing anomaly scores/investigation status, 8 invalid transaction IDs |

# 12. Data Cleaning Strategy

Based on the data-quality assessment, the identified issues will be
handled using appropriate cleaning and validation rules.

The original datasets will not be modified directly. Clean copies
will be created for further analysis and AI processing.

In [59]:
customers_clean = customers.copy()
transactions_clean = transactions.copy()
service_cases_clean = service_cases.copy()
fraud_alerts_clean = fraud_alerts.copy()

print("Clean copies created successfully.")

Clean copies created successfully.


## 12.1 Customers Data Cleaning

In [60]:
customers_clean["city"] = customers_clean["city"].fillna("Unknown")

customers_clean["kyc_status"] = customers_clean["kyc_status"].fillna("Unknown")

In [61]:
age_median = customers_clean["age"].median()

print("Median age:", age_median)

Median age: 45.0


In [62]:
customers_clean["age"] = customers_clean["age"].fillna(age_median)

In [63]:
customers_clean.isnull().sum()

customer_id         0
full_name           0
age                 0
gender              0
city                0
account_type        0
customer_segment    0
kyc_status          0
onboarding_date     0
dtype: int64

In [64]:
print("Rows:", len(customers_clean))
print("Duplicate rows:", customers_clean.duplicated().sum())
print(
    "Duplicate customer IDs:",
    customers_clean["customer_id"].duplicated().sum()
)

Rows: 1200
Duplicate rows: 0
Duplicate customer IDs: 0


In [65]:
# Save the cleaned customers dataset
customers_clean.to_csv(
    "customers_cleaned.csv",
    index=False
)

print("Cleaned customers dataset saved successfully.")
print("File: customers_cleaned.csv")
print("Rows:", len(customers_clean))
print("Columns:", len(customers_clean.columns))

Cleaned customers dataset saved successfully.
File: customers_cleaned.csv
Rows: 1200
Columns: 9


In [66]:
# Reload and verify the saved customers dataset
customers_final = pd.read_csv("customers_cleaned.csv")

print("Reloaded rows:", len(customers_final))
print("Reloaded columns:", len(customers_final.columns))
print(
    "Duplicate customer IDs:",
    customers_final["customer_id"].duplicated().sum()
)
print(
    "Missing values:",
    customers_final.isnull().sum().sum()
)

Reloaded rows: 1200
Reloaded columns: 9
Duplicate customer IDs: 0
Missing values: 0


### Customers Cleaning Result

- Missing age values were filled using the median age of 45.
- Missing city values were replaced with "Unknown".
- Missing KYC status values were replaced with "Unknown".
- No duplicate records were found.
- No duplicate customer IDs were found.
- The dataset remains at 1,200 records.

In [67]:
customers_clean = customers.copy()

print("Original rows:", len(customers))
print("Clean copy rows:", len(customers_clean))

Original rows: 1200
Clean copy rows: 1200


In [68]:
age_median = customers_clean["age"].median()

print("Median age:", age_median)

Median age: 45.0


In [69]:
customers_clean["age_missing_flag"] = (
    customers_clean["age"].isna()
)

print(
    "Missing age:",
    customers_clean["age_missing_flag"].sum()
)

Missing age: 14


In [70]:
customers_clean["age"] = customers_clean["age"].fillna(
    age_median
)

In [71]:
customers_clean["city_missing_flag"] = (
    customers_clean["city"].isna()
)

In [72]:
customers_clean["city"] = customers_clean["city"].fillna(
    "Unknown"
)

In [73]:
customers_clean["kyc_status_missing_flag"] = (
    customers_clean["kyc_status"].isna()
)

In [74]:
customers_clean["kyc_status"] = customers_clean["kyc_status"].fillna(
    "Unknown"
)

## Customer Onboarding Date Validation

In [75]:
customers_clean["onboarding_date"] = pd.to_datetime(
    customers_clean["onboarding_date"],
    errors="coerce"
)

print(
    "Invalid onboarding dates:",
    customers_clean["onboarding_date"].isna().sum()
)

print(
    "Minimum onboarding date:",
    customers_clean["onboarding_date"].min()
)

print(
    "Maximum onboarding date:",
    customers_clean["onboarding_date"].max()
)

Invalid onboarding dates: 0
Minimum onboarding date: 2021-01-03 00:00:00
Maximum onboarding date: 2026-06-25 00:00:00


In [76]:
invalid_age = customers_clean[
    (customers_clean["age"] < 18) |
    (customers_clean["age"] > 100)
]

print("Invalid age records:", len(invalid_age))

Invalid age records: 0


In [77]:
print("Gender:")
print(customers_clean["gender"].value_counts(dropna=False))

print("\nAccount Type:")
print(customers_clean["account_type"].value_counts(dropna=False))

print("\nCustomer Segment:")
print(customers_clean["customer_segment"].value_counts(dropna=False))

print("\nKYC Status:")
print(customers_clean["kyc_status"].value_counts(dropna=False))

Gender:
gender
Male      611
Female    554
Other      35
Name: count, dtype: int64

Account Type:
account_type
Savings    1013
Current     187
Name: count, dtype: int64

Customer Segment:
customer_segment
Retail       752
Premium      244
SME          160
Corporate     44
Name: count, dtype: int64

KYC Status:
kyc_status
Verified    1005
Pending      120
Expired       57
Unknown       18
Name: count, dtype: int64


In [78]:
print("========== FINAL CUSTOMER QUALITY CHECK ==========")

print("Total rows:", len(customers_clean))

print(
    "Duplicate rows:",
    customers_clean.duplicated().sum()
)

print(
    "Duplicate customer IDs:",
    customers_clean["customer_id"].duplicated().sum()
)

print("\nMissing values:")
print(customers_clean.isnull().sum())

print("\nData types:")
print(customers_clean.dtypes)

print("\n========== CHECK COMPLETED ==========")

========== FINAL CUSTOMER QUALITY CHECK ==========
Total rows: 1200
Duplicate rows: 0
Duplicate customer IDs: 0

Missing values:
customer_id                0
full_name                  0
age                        0
gender                     0
city                       0
account_type               0
customer_segment           0
kyc_status                 0
onboarding_date            0
age_missing_flag           0
city_missing_flag          0
kyc_status_missing_flag    0
dtype: int64

Data types:
customer_id                        object
full_name                          object
age                               float64
gender                             object
city                               object
account_type                       object
customer_segment                   object
kyc_status                         object
onboarding_date            datetime64[ns]
age_missing_flag                     bool
city_missing_flag                    bool
kyc_status_missing_flag            

In [79]:
customers_clean.to_csv(
    "customers_cleaned.csv",
    index=False
)

print("Cleaned customers dataset saved successfully.")
print("File: customers_cleaned.csv")
print("Rows:", len(customers_clean))
print("Columns:", len(customers_clean.columns))

Cleaned customers dataset saved successfully.
File: customers_cleaned.csv
Rows: 1200
Columns: 12


In [80]:
customers_final = pd.read_csv(
    "customers_cleaned.csv"
)

print("Reloaded rows:", len(customers_final))
print("Reloaded columns:", len(customers_final.columns))

print(
    "Duplicate customer IDs:",
    customers_final["customer_id"].duplicated().sum()
)

print(
    "Total missing values:",
    customers_final.isnull().sum().sum()
)

Reloaded rows: 1200
Reloaded columns: 12
Duplicate customer IDs: 0
Total missing values: 0


## 12.2 Transactions Data Cleaning

Before removing duplicate transaction records, duplicate transaction IDs
are inspected to determine whether they are exact duplicate records or
contain conflicting information.

In [81]:
duplicate_transactions = transactions_clean[
    transactions_clean["transaction_id"].duplicated(keep=False)
].sort_values("transaction_id")

duplicate_transactions

,transaction_id,customer_id,transaction_timestamp,transaction_type,amount,merchant_category,payment_channel,transaction_status,is_international,failed_attempts
1070,TXN0001071,CUST00920,2026-01-03 12:00:00,Credit,224.55,Investment,Net Banking,Success,0,0
7013,TXN0001071,CUST00920,2026-01-03 12:00:00,Credit,224.55,Investment,Net Banking,Success,0,0
1528,TXN0001529,CUST00460,2026-05-20 04:00:00,Transfer,4148.45,Groceries,UPI,Success,0,0
7006,TXN0001529,CUST00460,2026-05-20 04:00:00,Transfer,4148.45,Groceries,UPI,Success,0,0
1651,TXN0001652,CUST00742,2025-09-20 13:00:00,Debit,2117.29,Groceries,UPI,Pending,0,0
7005,TXN0001652,CUST00742,2025-09-20 13:00:00,Debit,2117.29,Groceries,UPI,Pending,0,0
1686,TXN0001687,CUST00241,2025-02-10 21:00:00,Debit,1487.75,Groceries,UPI,Success,0,0
7002,TXN0001687,CUST00241,2025-02-10 21:00:00,Debit,1487.75,Groceries,UPI,Success,0,0
2305,TXN0002306,CUST00007,2026-03-04 08:00:00,Debit,760.13,Food,Net Banking,Success,0,1
7000,TXN0002306,CUST00007,2026-03-04 08:00:00,Debit,760.13,Food,Net Banking,Success,0,1


In [82]:
print("Duplicate transaction records:", len(duplicate_transactions))
print(
    "Unique duplicated transaction IDs:",
    duplicate_transactions["transaction_id"].nunique()
)

Duplicate transaction records: 30
Unique duplicated transaction IDs: 15


### 12.2.1 Duplicate Transaction Removal

The transaction dataset contains 30 duplicate records belonging to
15 duplicated transaction IDs.

The duplicated records were inspected and found to contain identical
transaction information. Therefore, one copy of each duplicate record
will be retained and the extra duplicate records will be removed.

Original records: 7,015
Duplicate records to remove: 15
Expected cleaned records: 7,000

In [83]:
transactions_clean = transactions_clean.drop_duplicates(
    subset="transaction_id",
    keep="first"
)

In [84]:
print("Rows after duplicate removal:", len(transactions_clean))
print(
    "Duplicate transaction IDs:",
    transactions_clean["transaction_id"].duplicated().sum()
)

Rows after duplicate removal: 7000
Duplicate transaction IDs: 0


In [85]:
print(
    "Duplicate complete rows:",
    transactions_clean.duplicated().sum()
)

Duplicate complete rows: 0


In [86]:
negative_transactions_clean = transactions_clean[
    transactions_clean["amount"] < 0
]

print(
    "Negative transaction records:",
    len(negative_transactions_clean)
)

negative_transactions_clean[
    [
        "transaction_id",
        "customer_id",
        "transaction_type",
        "amount",
        "merchant_category",
        "payment_channel",
        "transaction_status"
    ]
]

Negative transaction records: 12


,transaction_id,customer_id,transaction_type,amount,merchant_category,payment_channel,transaction_status
212,TXN0000213,CUST01043,Debit,-1.0,Investment,Net Banking,Success
320,TXN0000321,CUST00854,Debit,-100.0,Food,UPI,Success
803,TXN0000804,CUST01122,Debit,-500.0,Travel,ATM,Success
2259,TXN0002260,CUST00586,Debit,-500.0,Entertainment,NaN,Success
3536,TXN0003537,CUST01053,Credit,-1.0,Travel,UPI,Success
3588,TXN0003589,CUST01200,Debit,-1.0,Entertainment,Credit Card,Success
4031,TXN0004032,CUST01124,Debit,-1.0,Shopping,UPI,Success
5835,TXN0005836,CUST01118,Credit,-100.0,Healthcare,UPI,Success
5954,TXN0005955,CUST01024,Credit,-500.0,Healthcare,Credit Card,Success
6406,TXN0006407,CUST00047,Credit,-1.0,Healthcare,Credit Card,Success


### 12.2.2 Negative Transaction Amount Validation

The cleaned transactions dataset contains 12 transactions with negative
amount values.

Negative amounts are not automatically removed because they may represent
legitimate financial events such as refunds, reversals, or corrections.

Since the current dataset does not contain a dedicated field explaining
the reason for the negative amount, these records are flagged for further
investigation rather than modified or deleted.

In [87]:
negative_transactions = transactions_clean[
    transactions_clean["amount"] < 0
].copy()

print("Negative transaction records:", len(negative_transactions))

Negative transaction records: 12


In [88]:
transactions_clean["negative_amount_flag"] = (
    transactions_clean["amount"] < 0
)

print(
    "Transactions flagged for negative amount:",
    transactions_clean["negative_amount_flag"].sum()
)

Transactions flagged for negative amount: 12


In [89]:
zero_transactions = transactions_clean[
    transactions_clean["amount"] == 0
]

print("Zero amount transactions:", len(zero_transactions))

Zero amount transactions: 6


### 12.2.3 Zero Transaction Amount Validation

The cleaned transactions dataset contains 6 transactions with an amount
of zero.

Zero-value transactions are inspected before cleaning because they may
represent legitimate events such as authorization attempts or system
records. They should not be removed without understanding their context.

In [90]:
zero_transactions = transactions_clean[
    transactions_clean["amount"] == 0
].copy()

print("Zero amount transactions:", len(zero_transactions))

zero_transactions[
    [
        "transaction_id",
        "customer_id",
        "transaction_type",
        "amount",
        "merchant_category",
        "payment_channel",
        "transaction_status",
        "failed_attempts"
    ]
]

Zero amount transactions: 6


,transaction_id,customer_id,transaction_type,amount,merchant_category,payment_channel,transaction_status,failed_attempts
1208,TXN0001209,CUST00792,Transfer,0.0,Shopping,UPI,Failed,0
1349,TXN0001350,CUST01017,Credit,0.0,Investment,UPI,Success,0
1401,TXN0001402,CUST00894,Debit,0.0,Shopping,Credit Card,Success,1
1779,TXN0001780,CUST00574,Debit,0.0,Salary,Net Banking,Success,2
6334,TXN0006335,CUST00373,Debit,0.0,Rent,UPI,Success,0
6440,TXN0006441,CUST01182,Credit,0.0,Healthcare,Debit Card,Success,0


In [91]:
zero_transactions["transaction_status"].value_counts(dropna=False)

transaction_status
Success    5
Failed     1
Name: count, dtype: int64

In [92]:
transactions_clean["zero_amount_flag"] = (
    transactions_clean["amount"] == 0
)

print(
    "Transactions flagged for zero amount:",
    transactions_clean["zero_amount_flag"].sum()
)

Transactions flagged for zero amount: 6


In [93]:
transactions_clean.isnull().sum()

transaction_id            0
customer_id              42
transaction_timestamp     0
transaction_type          0
amount                   56
merchant_category        70
payment_channel          84
transaction_status       70
is_international          0
failed_attempts           0
negative_amount_flag      0
zero_amount_flag          0
dtype: int64

In [94]:
missing_transactions_clean = pd.DataFrame({
    "missing_count": transactions_clean.isnull().sum(),
    "missing_percentage": transactions_clean.isnull().mean() * 100
})

missing_transactions_clean

,missing_count,missing_percentage
transaction_id,0,0.0
customer_id,42,0.6
transaction_timestamp,0,0.0
transaction_type,0,0.0
amount,56,0.8
merchant_category,70,1.0
payment_channel,84,1.2
transaction_status,70,1.0
is_international,0,0.0
failed_attempts,0,0.0


### 12.2.4 Missing Customer IDs

There are 42 transactions with missing customer IDs (0.60%).

Customer IDs are important for linking transactions with customer records.
Because the correct customer cannot be reliably inferred from the available
transaction fields, missing customer IDs are not replaced with artificial
customer IDs.

These records are retained and flagged for further investigation.

In [95]:
transactions_clean["customer_id_missing_flag"] = (
    transactions_clean["customer_id"].isna()
)

print(
    "Missing customer IDs:",
    transactions_clean["customer_id_missing_flag"].sum()
)

Missing customer IDs: 42


In [96]:
amount_median = transactions_clean["amount"].median()

print("Transaction amount median:", amount_median)

Transaction amount median: 1096.2649999999999


In [97]:
transactions_clean["amount_missing_flag"] = (
    transactions_clean["amount"].isna()
)

transactions_clean["amount"] = transactions_clean["amount"].fillna(
    amount_median
)

print(
    "Missing amounts after cleaning:",
    transactions_clean["amount"].isna().sum()
)

Missing amounts after cleaning: 0


In [98]:
categorical_missing_columns = [
    "merchant_category",
    "payment_channel",
    "transaction_status"
]

for column in categorical_missing_columns:
    transactions_clean[column] = transactions_clean[column].fillna("Unknown")

In [99]:
transactions_clean[
    [
        "merchant_category",
        "payment_channel",
        "transaction_status"
    ]
].isna().sum()

merchant_category     0
payment_channel       0
transaction_status    0
dtype: int64

In [100]:
transactions_clean.isnull().sum()

transaction_id               0
customer_id                 42
transaction_timestamp        0
transaction_type             0
amount                       0
merchant_category            0
payment_channel              0
transaction_status           0
is_international             0
failed_attempts              0
negative_amount_flag         0
zero_amount_flag             0
customer_id_missing_flag     0
amount_missing_flag          0
dtype: int64

In [101]:
print("Final transaction rows:", len(transactions_clean))
print(
    "Duplicate transaction IDs:",
    transactions_clean["transaction_id"].duplicated().sum()
)

Final transaction rows: 7000
Duplicate transaction IDs: 0


## 12.3 Transactions Categorical Validation

Categorical fields are checked after missing-value handling to identify
unexpected, inconsistent, or invalid category values.

The validation focuses on transaction type, merchant category,
payment channel, and transaction status.

In [102]:
transactions_clean["transaction_type"].value_counts(dropna=False)

transaction_type
Debit       3492
Transfer    2095
Credit      1413
Name: count, dtype: int64

In [103]:
transactions_clean["merchant_category"].value_counts(dropna=False)

merchant_category
Travel           738
Entertainment    730
Salary           715
Shopping         711
Groceries        711
Investment       707
Rent             670
Food             667
Utilities        644
Healthcare       637
Unknown           70
Name: count, dtype: int64

In [104]:
transactions_clean["payment_channel"].value_counts(dropna=False)

payment_channel
UPI            2968
Debit Card     1385
Net Banking     951
Credit Card     877
ATM             719
Unknown          84
upi              12
credit_card       2
debit_card        2
Name: count, dtype: int64

In [105]:
transactions_clean["transaction_status"].value_counts(dropna=False)

transaction_status
Success    6019
Failed      561
Pending     350
Unknown      70
Name: count, dtype: int64

In [106]:
transactions_clean["is_international"].value_counts(dropna=False)

is_international
0    6594
1     406
Name: count, dtype: int64

In [107]:
transactions_clean["failed_attempts"].describe()

count    7000.000000
mean        0.250429
std         0.497744
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         3.000000
Name: failed_attempts, dtype: float64

### 12.3.1 Payment Channel Standardization

Payment channel values are standardized to maintain consistent
categorical representation.

Equivalent values such as `upi`, `credit_card`, and `debit_card`
are converted to the standard naming convention used in the dataset.

In [108]:
transactions_clean["payment_channel"] = (
    transactions_clean["payment_channel"]
    .str.strip()
    .replace({
        "upi": "UPI",
        "credit_card": "Credit Card",
        "debit_card": "Debit Card"
    })
)

In [109]:
transactions_clean["payment_channel"].value_counts(dropna=False)

payment_channel
UPI            2980
Debit Card     1387
Net Banking     951
Credit Card     879
ATM             719
Unknown          84
Name: count, dtype: int64

In [110]:
transactions_clean["payment_channel"].unique()

array(['Net Banking', 'Credit Card', 'Debit Card', 'ATM', 'UPI',
       'Unknown'], dtype=object)

## 12.4 Transaction Date & Time Validation

Transaction timestamps are validated to ensure that the field is
stored in the correct datetime format and does not contain invalid
or future transaction dates.

The validation checks:
- Datatype of transaction_timestamp
- Missing timestamps
- Invalid datetime values
- Future transaction dates
- Minimum and maximum transaction dates

In [111]:
print("Transaction timestamp datatype:")
print(transactions_clean["transaction_timestamp"].dtype)

Transaction timestamp datatype:
object


In [112]:
transactions_clean["transaction_timestamp"] = pd.to_datetime(
    transactions_clean["transaction_timestamp"],
    errors="coerce"
)

print("Invalid timestamps:", transactions_clean["transaction_timestamp"].isna().sum())

Invalid timestamps: 0


In [113]:
print("Minimum transaction timestamp:",
      transactions_clean["transaction_timestamp"].min())

print("Maximum transaction timestamp:",
      transactions_clean["transaction_timestamp"].max())

Minimum transaction timestamp: 2025-01-01 02:00:00
Maximum transaction timestamp: 2026-09-30 00:00:00


In [114]:
future_transactions = transactions_clean[
    transactions_clean["transaction_timestamp"] > pd.Timestamp.now()
]

print("Future transaction records:", len(future_transactions))

Future transaction records: 0


In [115]:
future_transactions[
    [
        "transaction_id",
        "customer_id",
        "transaction_timestamp",
        "amount",
        "transaction_status"
    ]
].head(20)

,transaction_id,customer_id,transaction_timestamp,amount,transaction_status


### 12.5 Numeric Data Validation

Numeric transaction fields are validated to identify invalid,
unexpected, or extreme values after missing-value handling.

The validation focuses on:
- Transaction amount
- Failed attempts
- Minimum and maximum values
- Statistical distribution

In [116]:
print("Transaction Amount Statistics:")
print(transactions_clean["amount"].describe())

Transaction Amount Statistics:
count     7000.000000
mean      1891.586379
std       2616.426634
min       -500.000000
25%        537.345000
50%       1096.265000
75%       2226.677500
max      43312.140000
Name: amount, dtype: float64


In [117]:
print("Failed Attempts Statistics:")
print(transactions_clean["failed_attempts"].describe())

Failed Attempts Statistics:
count    7000.000000
mean        0.250429
std         0.497744
min         0.000000
25%         0.000000
50%         0.000000
75%         0.000000
max         3.000000
Name: failed_attempts, dtype: float64


In [118]:
invalid_failed_attempts = transactions_clean[
    transactions_clean["failed_attempts"] < 0
]

print(
    "Negative failed attempts:",
    len(invalid_failed_attempts)
)

Negative failed attempts: 0


In [119]:
high_amount_transactions = transactions_clean[
    transactions_clean["amount"] > transactions_clean["amount"].quantile(0.99)
]

print(
    "Transactions above 99th percentile:",
    len(high_amount_transactions)
)

high_amount_transactions[
    [
        "transaction_id",
        "customer_id",
        "amount",
        "merchant_category",
        "transaction_status"
    ]
].head(20)

Transactions above 99th percentile: 70


,transaction_id,customer_id,amount,merchant_category,transaction_status
7,TXN0000008,CUST00144,18402.82,Rent,Success
312,TXN0000313,CUST01184,21543.87,Food,Success
372,TXN0000373,CUST00219,13203.56,Travel,Success
455,TXN0000456,CUST01119,30518.63,Entertainment,Success
578,TXN0000579,CUST00195,23858.41,Unknown,Success
886,TXN0000887,CUST00557,31668.71,Travel,Success
1136,TXN0001137,CUST00980,14903.46,Food,Success
1160,TXN0001161,CUST01008,14740.26,Entertainment,Success
1195,TXN0001196,CUST00072,16635.89,Investment,Success
1225,TXN0001226,CUST00064,15235.54,Groceries,Success


### 12.6 Transaction Amount Outlier Flagging

High transaction amounts are identified using the 99th percentile
as a statistical outlier threshold.

These transactions are not removed because high-value transactions
may represent legitimate financial activity and can also be useful
for later fraud and anomaly detection.

An outlier flag is therefore created instead of deleting the records.

In [120]:
amount_99th_percentile = transactions_clean["amount"].quantile(0.99)

print("99th percentile amount:", amount_99th_percentile)

99th percentile amount: 12135.41200000002


In [121]:
transactions_clean["high_amount_flag"] = (
    transactions_clean["amount"] > amount_99th_percentile
)

print(
    "High amount transactions:",
    transactions_clean["high_amount_flag"].sum()
)

High amount transactions: 70


In [122]:
transactions_clean[
    transactions_clean["high_amount_flag"]
][
    [
        "transaction_id",
        "customer_id",
        "amount",
        "merchant_category",
        "transaction_status",
        "high_amount_flag"
    ]
].head(20)

,transaction_id,customer_id,amount,merchant_category,transaction_status,high_amount_flag
7,TXN0000008,CUST00144,18402.82,Rent,Success,True
312,TXN0000313,CUST01184,21543.87,Food,Success,True
372,TXN0000373,CUST00219,13203.56,Travel,Success,True
455,TXN0000456,CUST01119,30518.63,Entertainment,Success,True
578,TXN0000579,CUST00195,23858.41,Unknown,Success,True
886,TXN0000887,CUST00557,31668.71,Travel,Success,True
1136,TXN0001137,CUST00980,14903.46,Food,Success,True
1160,TXN0001161,CUST01008,14740.26,Entertainment,Success,True
1195,TXN0001196,CUST00072,16635.89,Investment,Success,True
1225,TXN0001226,CUST00064,15235.54,Groceries,Success,True


### 12.7 Final Transaction Data Quality Check

A final quality check is performed after cleaning and standardization
to confirm that the transactions dataset is ready for downstream
analytics, machine learning, fraud detection, and financial insights.

The final validation checks:
- Total records
- Duplicate transaction IDs
- Missing values
- Standardized categorical values
- Referential integrity
- Data types
- Cleaning flags

In [123]:
print("Final transaction rows:", len(transactions_clean))

print(
    "Duplicate transaction IDs:",
    transactions_clean["transaction_id"].duplicated().sum()
)

Final transaction rows: 7000
Duplicate transaction IDs: 0


In [124]:
final_missing = pd.DataFrame({
    "missing_count": transactions_clean.isnull().sum(),
    "missing_percentage": transactions_clean.isnull().mean() * 100
})

final_missing

,missing_count,missing_percentage
transaction_id,0,0.0
customer_id,42,0.6
transaction_timestamp,0,0.0
transaction_type,0,0.0
amount,0,0.0
merchant_category,0,0.0
payment_channel,0,0.0
transaction_status,0,0.0
is_international,0,0.0
failed_attempts,0,0.0


In [125]:
print("Payment channels:")
print(transactions_clean["payment_channel"].unique())

print("\nTransaction statuses:")
print(transactions_clean["transaction_status"].unique())

print("\nTransaction types:")
print(transactions_clean["transaction_type"].unique())

Payment channels:
['Net Banking' 'Credit Card' 'Debit Card' 'ATM' 'UPI' 'Unknown']

Transaction statuses:
['Failed' 'Success' 'Pending' 'Unknown']

Transaction types:
['Credit' 'Transfer' 'Debit']


In [126]:
print("Negative amount flags:",
      transactions_clean["negative_amount_flag"].sum())

print("Zero amount flags:",
      transactions_clean["zero_amount_flag"].sum())

print("Missing customer ID flags:",
      transactions_clean["customer_id_missing_flag"].sum())

print("Missing amount flags:",
      transactions_clean["amount_missing_flag"].sum())

print("High amount flags:",
      transactions_clean["high_amount_flag"].sum())

Negative amount flags: 12
Zero amount flags: 6
Missing customer ID flags: 42
Missing amount flags: 56
High amount flags: 70


In [127]:
transactions_clean.dtypes

transaction_id                      object
customer_id                         object
transaction_timestamp       datetime64[ns]
transaction_type                    object
amount                             float64
merchant_category                   object
payment_channel                     object
transaction_status                  object
is_international                     int64
failed_attempts                      int64
negative_amount_flag                  bool
zero_amount_flag                      bool
customer_id_missing_flag              bool
amount_missing_flag                   bool
high_amount_flag                      bool
dtype: object

## 12.8 Export Cleaned Transactions Dataset

The validated transactions dataset is exported as a separate cleaned file.
The original raw dataset is preserved unchanged for reproducibility and audit purposes.

In [128]:
# Save the cleaned transactions dataset
transactions_clean.to_csv(
    "transactions_cleaned.csv",
    index=False
)

print("Cleaned transactions dataset saved successfully.")
print("File: transactions_cleaned.csv")
print("Rows:", len(transactions_clean))
print("Columns:", len(transactions_clean.columns))

Cleaned transactions dataset saved successfully.
File: transactions_cleaned.csv
Rows: 7000
Columns: 15


In [129]:
# Reload the saved dataset and verify
transactions_final = pd.read_csv("transactions_cleaned.csv")

print("Reloaded rows:", len(transactions_final))
print("Reloaded columns:", len(transactions_final.columns))
print("Duplicate transaction IDs:",
      transactions_final["transaction_id"].duplicated().sum())

Reloaded rows: 7000
Reloaded columns: 15
Duplicate transaction IDs: 0


## 13.1 Fraud Alerts Data Cleaning

A separate copy of the fraud alerts dataset is created for cleaning.
The original fraud alerts dataset is preserved unchanged for reproducibility and audit purposes.

In [130]:
fraud_alerts_clean = fraud_alerts.copy()

print("Clean copy created successfully.")
print("Original rows:", len(fraud_alerts))
print("Clean copy rows:", len(fraud_alerts_clean))

Clean copy created successfully.
Original rows: 1100
Clean copy rows: 1100


In [131]:
anomaly_median = fraud_alerts_clean["anomaly_score"].median()

print("Median anomaly score:", anomaly_median)

Median anomaly score: 0.345


In [132]:
fraud_alerts_clean["anomaly_score_missing_flag"] = (
    fraud_alerts_clean["anomaly_score"].isna()
)

print(
    "Missing anomaly scores:",
    fraud_alerts_clean["anomaly_score_missing_flag"].sum()
)

Missing anomaly scores: 16


In [133]:
fraud_alerts_clean["anomaly_score"] = (
    fraud_alerts_clean["anomaly_score"]
    .fillna(anomaly_median)
)

print(
    "Remaining missing anomaly scores:",
    fraud_alerts_clean["anomaly_score"].isna().sum()
)

Remaining missing anomaly scores: 0


In [134]:
invalid_anomaly_scores = fraud_alerts_clean[
    (fraud_alerts_clean["anomaly_score"] < 0) |
    (fraud_alerts_clean["anomaly_score"] > 1)
]

print(
    "Invalid anomaly scores:",
    len(invalid_anomaly_scores)
)

Invalid anomaly scores: 0


In [135]:
print(fraud_alerts_clean["anomaly_score"].describe())

count    1100.000000
mean        0.416876
std         0.270172
min         0.008000
25%         0.196000
50%         0.345000
75%         0.655000
max         0.990000
Name: anomaly_score, dtype: float64


In [136]:
fraud_alerts_clean["investigation_status_missing_flag"] = (
    fraud_alerts_clean["investigation_status"].isna()
)

print(
    "Missing investigation status:",
    fraud_alerts_clean[
        "investigation_status_missing_flag"
    ].sum()
)

Missing investigation status: 13


In [137]:
fraud_alerts_clean["investigation_status"] = (
    fraud_alerts_clean["investigation_status"]
    .fillna("Unknown")
)

print(
    "Remaining missing investigation status:",
    fraud_alerts_clean["investigation_status"].isna().sum()
)

Remaining missing investigation status: 0


In [138]:
fraud_alerts_clean["transaction_id_invalid_flag"] = (
    ~fraud_alerts_clean["transaction_id"].isin(
        transactions_clean["transaction_id"]
    )
)

print(
    "Invalid transaction references:",
    fraud_alerts_clean[
        "transaction_id_invalid_flag"
    ].sum()
)

Invalid transaction references: 8


In [139]:
invalid_alert_transactions = fraud_alerts_clean[
    fraud_alerts_clean["transaction_id_invalid_flag"]
]

invalid_alert_transactions[
    [
        "alert_id",
        "transaction_id",
        "anomaly_score",
        "risk_level",
        "investigation_status"
    ]
]

,alert_id,transaction_id,anomaly_score,risk_level,investigation_status
238,ALRT000239,TXN99995,0.218,Low,Cleared
239,ALRT000240,TXN99990,0.248,Medium,Cleared
258,ALRT000259,TXN99997,0.979,High,Under Review
287,ALRT000288,TXN99993,0.260,Low,Cleared
391,ALRT000392,TXN99994,0.667,High,Under Review
1000,ALRT001001,TXN99996,0.276,Low,Cleared
1001,ALRT001002,TXN99991,0.272,Medium,Under Review
1006,ALRT001007,TXN99992,0.122,Low,Cleared


In [140]:
print("Alert types:")
print(fraud_alerts_clean["alert_type"].value_counts(dropna=False))

print("\nRisk levels:")
print(fraud_alerts_clean["risk_level"].value_counts(dropna=False))

print("\nInvestigation statuses:")
print(
    fraud_alerts_clean[
        "investigation_status"
    ].value_counts(dropna=False)
)

print("\nAnalyst actions:")
print(
    fraud_alerts_clean[
        "analyst_action"
    ].value_counts(dropna=False)
)

Alert types:
alert_type
Rapid Transactions        232
Unusual Location          230
International Activity    214
Multiple Failures         212
High Amount               212
Name: count, dtype: int64

Risk levels:
risk_level
Low         469
Medium      366
High        213
Critical     52
Name: count, dtype: int64

Investigation statuses:
investigation_status
Cleared         401
Under Review    341
New             266
Confirmed        79
Unknown          13
Name: count, dtype: int64

Analyst actions:
analyst_action
No Action            296
Escalate             294
Block Transaction    268
Contact Customer     242
Name: count, dtype: int64


## 13.7 Fraud Alert Timestamp Validation

Alert timestamps are validated to ensure that they are stored in the correct
datetime format and do not contain missing, invalid, or future dates.

In [141]:
print("Alert timestamp datatype:")
print(fraud_alerts_clean["alert_timestamp"].dtype)

Alert timestamp datatype:
object


In [142]:
fraud_alerts_clean["alert_timestamp"] = pd.to_datetime(
    fraud_alerts_clean["alert_timestamp"],
    errors="coerce"
)

print(
    "Invalid alert timestamps:",
    fraud_alerts_clean["alert_timestamp"].isna().sum()
)

Invalid alert timestamps: 0


In [143]:
print(
    "Minimum alert timestamp:",
    fraud_alerts_clean["alert_timestamp"].min()
)

print(
    "Maximum alert timestamp:",
    fraud_alerts_clean["alert_timestamp"].max()
)

Minimum alert timestamp: 2025-01-01 00:00:00
Maximum alert timestamp: 2026-09-30 00:00:00


In [144]:
future_alerts = fraud_alerts_clean[
    fraud_alerts_clean["alert_timestamp"] > pd.Timestamp.now()
]

print(
    "Future alert records:",
    len(future_alerts)
)

Future alert records: 0


## 13.8 Risk and Anomaly Consistency Validation

The anomaly score and risk level are compared to identify potentially
inconsistent fraud-alert records. These records are flagged for review
rather than automatically modified.

In [145]:
fraud_alerts_clean.groupby("risk_level")["anomaly_score"].agg(
    ["count", "min", "mean", "max"]
)

,count,min,mean,max
risk_level,,,,
Critical,52,0.345,0.821327,0.981
High,213,0.345,0.818972,0.990
Low,469,0.008,0.285923,0.882
Medium,366,0.020,0.293213,0.779


In [146]:
risk_anomaly_check = fraud_alerts_clean[
    [
        "alert_id",
        "anomaly_score",
        "risk_level",
        "investigation_status"
    ]
].copy()

risk_anomaly_check

,alert_id,anomaly_score,risk_level,investigation_status
0,ALRT000001,0.194,Low,New
1,ALRT000002,0.084,Low,Under Review
2,ALRT000003,0.862,Critical,Under Review
3,ALRT000004,0.248,Low,Under Review
4,ALRT000005,0.164,Low,Cleared
...,...,...,...,...
1095,ALRT001096,0.368,Medium,New
1096,ALRT001097,0.156,Medium,Unknown
1097,ALRT001098,0.357,Low,New
1098,ALRT001099,0.232,Medium,Confirmed


In [147]:
print("Risk level distribution:")
print(
    fraud_alerts_clean["risk_level"].value_counts()
)

print("\nAverage anomaly score by risk level:")
print(
    fraud_alerts_clean.groupby("risk_level")["anomaly_score"].mean()
)

Risk level distribution:
risk_level
Low         469
Medium      366
High        213
Critical     52
Name: count, dtype: int64

Average anomaly score by risk level:
risk_level
Critical    0.821327
High        0.818972
Low         0.285923
Medium      0.293213
Name: anomaly_score, dtype: float64


## 13.8.1 Individual Risk-Anomaly Consistency Check

Individual fraud alerts are checked for potentially inconsistent combinations
between anomaly score and assigned risk level. Inconsistent records are
flagged for review rather than automatically modified.

In [148]:
# Define expected anomaly-score ranges for risk levels
fraud_alerts_clean["risk_anomaly_inconsistency_flag"] = (
    ((fraud_alerts_clean["risk_level"] == "Low") &
     (fraud_alerts_clean["anomaly_score"] >= 0.70))
    |
    ((fraud_alerts_clean["risk_level"] == "Medium") &
     (fraud_alerts_clean["anomaly_score"] >= 0.80))
    |
    ((fraud_alerts_clean["risk_level"] == "High") &
     (fraud_alerts_clean["anomaly_score"] < 0.50))
    |
    ((fraud_alerts_clean["risk_level"] == "Critical") &
     (fraud_alerts_clean["anomaly_score"] < 0.70))
)

print(
    "Risk-anomaly inconsistent records:",
    fraud_alerts_clean["risk_anomaly_inconsistency_flag"].sum()
)

Risk-anomaly inconsistent records: 12


In [149]:
risk_anomaly_issues = fraud_alerts_clean[
    fraud_alerts_clean["risk_anomaly_inconsistency_flag"]
]

risk_anomaly_issues[
    [
        "alert_id",
        "anomaly_score",
        "risk_level",
        "investigation_status"
    ]
].head(20)

,alert_id,anomaly_score,risk_level,investigation_status
205,ALRT000206,0.694,Critical,Cleared
457,ALRT000458,0.736,Low,Cleared
589,ALRT000590,0.882,Low,Confirmed
619,ALRT000620,0.746,Low,Confirmed
737,ALRT000738,0.660,Critical,Cleared
757,ALRT000758,0.789,Low,New
813,ALRT000814,0.687,Critical,Cleared
819,ALRT000820,0.345,High,New
830,ALRT000831,0.677,Critical,New
937,ALRT000938,0.345,Critical,Under Review


## 13.9 Final Fraud Alerts Data Quality Check

A final quality check is performed after cleaning and validation to confirm
that the fraud alerts dataset is ready for downstream fraud analysis,
anomaly detection, investigation workflows, and operational dashboards.

The final checks include:
- Total records
- Duplicate alert IDs
- Missing values
- Invalid anomaly scores
- Invalid transaction references
- Invalid timestamps
- Risk-anomaly consistency flags
- Data types
- Cleaning flags

In [150]:
print("========== FINAL FRAUD ALERTS QUALITY CHECK ==========")

# 1. Total records
print("Total fraud alert records:", len(fraud_alerts_clean))

# 2. Duplicate alert IDs
print(
    "Duplicate alert IDs:",
    fraud_alerts_clean["alert_id"].duplicated().sum()
)

# 3. Missing values
print("\nMissing values:")
print(fraud_alerts_clean.isnull().sum())

# 4. Invalid anomaly scores
invalid_anomaly_final = fraud_alerts_clean[
    (fraud_alerts_clean["anomaly_score"] < 0) |
    (fraud_alerts_clean["anomaly_score"] > 1)
]

print(
    "\nInvalid anomaly scores:",
    len(invalid_anomaly_final)
)

# 5. Invalid transaction references
print(
    "Invalid transaction references:",
    fraud_alerts_clean["transaction_id_invalid_flag"].sum()
)

# 6. Invalid timestamps
print(
    "Invalid alert timestamps:",
    fraud_alerts_clean["alert_timestamp"].isnull().sum()
)

# 7. Risk-anomaly consistency
print(
    "Risk-anomaly inconsistency flags:",
    fraud_alerts_clean["risk_anomaly_inconsistency_flag"].sum()
)

# 8. Data types
print("\nData types:")
print(fraud_alerts_clean.dtypes)

print("\n========== QUALITY CHECK COMPLETED ==========")

========== FINAL FRAUD ALERTS QUALITY CHECK ==========
Total fraud alert records: 1100
Duplicate alert IDs: 0

Missing values:
alert_id                             0
transaction_id                       0
alert_timestamp                      0
anomaly_score                        0
alert_type                           0
risk_level                           0
investigation_status                 0
analyst_action                       0
anomaly_score_missing_flag           0
investigation_status_missing_flag    0
transaction_id_invalid_flag          0
risk_anomaly_inconsistency_flag      0
dtype: int64

Invalid anomaly scores: 0
Invalid transaction references: 8
Invalid alert timestamps: 0
Risk-anomaly inconsistency flags: 12

Data types:
alert_id                                     object
transaction_id                               object
alert_timestamp                      datetime64[ns]
anomaly_score                               float64
alert_type                                   

In [151]:
# Save the cleaned fraud alerts dataset

fraud_alerts_clean.to_csv(
    "fraud_alerts_cleaned.csv",
    index=False
)

print("Cleaned fraud alerts dataset saved successfully.")
print("File: fraud_alerts_cleaned.csv")
print("Rows:", len(fraud_alerts_clean))
print("Columns:", len(fraud_alerts_clean.columns))

Cleaned fraud alerts dataset saved successfully.
File: fraud_alerts_cleaned.csv
Rows: 1100
Columns: 12


In [152]:
# Reload the saved dataset and verify

fraud_alerts_final = pd.read_csv("fraud_alerts_cleaned.csv")

print("Reloaded rows:", len(fraud_alerts_final))
print("Reloaded columns:", len(fraud_alerts_final.columns))

print(
    "Duplicate alert IDs:",
    fraud_alerts_final["alert_id"].duplicated().sum()
)

Reloaded rows: 1100
Reloaded columns: 12
Duplicate alert IDs: 0


## 14. Service Cases Data Cleaning

The service cases dataset is cleaned and validated before being used
for customer service analytics, SLA analysis, case workflow automation,
and AI-assisted customer support.

The original dataset is preserved unchanged, while a separate cleaned
copy is created for downstream analysis.

In [153]:
# Create a separate copy for cleaning

service_cases_clean = service_cases.copy()

print("Original service case rows:", len(service_cases))
print("Clean service case rows:", len(service_cases_clean))

print("\nColumns:")
print(service_cases_clean.columns.tolist())

print("\nData types:")
print(service_cases_clean.dtypes)

print("\nMissing values:")
print(service_cases_clean.isnull().sum())

Original service case rows: 2500
Clean service case rows: 2500

Columns:
['case_id', 'customer_id', 'transaction_id', 'created_at', 'case_type', 'priority', 'status', 'resolution_time_hours', 'sla_breached', 'channel', 'customer_query']

Data types:
case_id                   object
customer_id               object
transaction_id            object
created_at                object
case_type                 object
priority                  object
status                    object
resolution_time_hours    float64
sla_breached                bool
channel                   object
customer_query            object
dtype: object

Missing values:
case_id                   0
customer_id               0
transaction_id            0
created_at                0
case_type                 0
priority                 37
status                    0
resolution_time_hours    50
sla_breached              0
channel                   0
customer_query           30
dtype: int64


In [154]:
print("Duplicate case IDs:",
      service_cases_clean["case_id"].duplicated().sum())

Duplicate case IDs: 0


In [155]:
# 14.1 Basic Service Cases Quality Check

print("========== SERVICE CASES BASIC CHECK ==========")

print("Total rows:", len(service_cases_clean))
print("Total columns:", len(service_cases_clean.columns))

print("\nColumns:")
print(service_cases_clean.columns.tolist())

print("\nData types:")
print(service_cases_clean.dtypes)

print("\nMissing values:")
print(service_cases_clean.isnull().sum())

print("\nDuplicate case IDs:")
print(service_cases_clean["case_id"].duplicated().sum())

print("\n========== CHECK COMPLETED ==========")

========== SERVICE CASES BASIC CHECK ==========
Total rows: 2500
Total columns: 11

Columns:
['case_id', 'customer_id', 'transaction_id', 'created_at', 'case_type', 'priority', 'status', 'resolution_time_hours', 'sla_breached', 'channel', 'customer_query']

Data types:
case_id                   object
customer_id               object
transaction_id            object
created_at                object
case_type                 object
priority                  object
status                    object
resolution_time_hours    float64
sla_breached                bool
channel                   object
customer_query            object
dtype: object

Missing values:
case_id                   0
customer_id               0
transaction_id            0
created_at                0
case_type                 0
priority                 37
status                    0
resolution_time_hours    50
sla_breached              0
channel                   0
customer_query           30
dtype: int64

Duplicate case

In [156]:
# 14.2 Categorical Value Validation

categorical_columns = [
    "case_type",
    "priority",
    "status",
    "channel",
    "sla_breached"
]

for column in categorical_columns:
    print(f"\n===== {column} =====")
    print(service_cases_clean[column].value_counts(dropna=False))


===== case_type =====
case_type
KYC                    396
Card Issue             366
Refund                 355
General Enquiry        353
Account Issue          350
Transaction Dispute    341
Payment Failure        339
Name: count, dtype: int64

===== priority =====
priority
Medium      1216
Low          591
High         546
Critical     110
NaN           37
Name: count, dtype: int64

===== status =====
status
Resolved       1286
In Progress     611
Open            418
Escalated       180
resolved          5
Name: count, dtype: int64

===== channel =====
channel
Chat      549
App       511
Email     505
Branch    471
Phone     464
Name: count, dtype: int64

===== sla_breached =====
sla_breached
False    1835
True      665
Name: count, dtype: int64


In [157]:
# 14.3 Resolution Time Validation

print("Resolution time statistics:")
print(service_cases_clean["resolution_time_hours"].describe())

print(
    "\nNegative resolution times:",
    (service_cases_clean["resolution_time_hours"] < 0).sum()
)

Resolution time statistics:
count    2450.000000
mean       18.434984
std        12.172637
min        -9.000000
25%         9.392500
50%        16.230000
75%        24.677500
max       103.910000
Name: resolution_time_hours, dtype: float64

Negative resolution times: 12


In [158]:
# 14.4 Service Case Timestamp Validation

print("Created_at datatype before conversion:")
print(service_cases_clean["created_at"].dtype)

service_cases_clean["created_at"] = pd.to_datetime(
    service_cases_clean["created_at"],
    errors="coerce"
)

print("\nInvalid created_at values:")
print(service_cases_clean["created_at"].isnull().sum())

print("\nMinimum created_at:")
print(service_cases_clean["created_at"].min())

print("\nMaximum created_at:")
print(service_cases_clean["created_at"].max())

Created_at datatype before conversion:
object

Invalid created_at values:
0

Minimum created_at:
2025-01-01 00:00:00

Maximum created_at:
2026-09-30 00:00:00


In [159]:
# 14.5.1 Check missing resolution time against case status

missing_resolution = service_cases_clean[
    service_cases_clean["resolution_time_hours"].isna()
].copy()

print("Cases with missing resolution time:", len(missing_resolution))

print("\nStatus of cases with missing resolution time:")
print(
    missing_resolution["status"]
    .value_counts(dropna=False)
)

Cases with missing resolution time: 50

Status of cases with missing resolution time:
status
In Progress    19
Resolved       17
Escalated       7
Open            7
Name: count, dtype: int64


In [160]:
# Check negative resolution time records

negative_resolution = service_cases_clean[
    service_cases_clean["resolution_time_hours"] < 0
].copy()

print("Negative resolution time records:", len(negative_resolution))

negative_resolution[
    [
        "case_id",
        "status",
        "resolution_time_hours",
        "sla_breached"
    ]
]

Negative resolution time records: 12


,case_id,status,resolution_time_hours,sla_breached
544,CASE000545,In Progress,-5.0,False
639,CASE000640,In Progress,-8.0,True
1099,CASE001100,Escalated,-6.0,False
1159,CASE001160,Resolved,-5.0,False
1161,CASE001162,Open,-9.0,False
1729,CASE001730,In Progress,-8.0,False
1752,CASE001753,Resolved,-6.0,False
1792,CASE001793,Open,-5.0,False
1823,CASE001824,In Progress,-8.0,True
2100,CASE002101,Open,-1.0,True


In [161]:
# 14.5.2 Standardize case status values

service_cases_clean["status"] = (
    service_cases_clean["status"]
    .astype("string")
    .str.strip()
)

service_cases_clean["status"] = (
    service_cases_clean["status"]
    .replace({
        "resolved": "Resolved"
    })
)

print(service_cases_clean["status"].value_counts(dropna=False))

status
Resolved       1291
In Progress     611
Open            418
Escalated       180
Name: count, dtype: Int64


In [162]:
# Missing priority flag

service_cases_clean["priority_missing_flag"] = (
    service_cases_clean["priority"].isna()
)

print(
    "Missing priority records:",
    service_cases_clean["priority_missing_flag"].sum()
)

Missing priority records: 37


In [163]:
service_cases_clean["priority"] = (
    service_cases_clean["priority"]
    .fillna("Unknown")
)

print(service_cases_clean["priority"].value_counts(dropna=False))

priority
Medium      1216
Low          591
High         546
Critical     110
Unknown       37
Name: count, dtype: int64


In [164]:
# Missing customer query flag

service_cases_clean["customer_query_missing_flag"] = (
    service_cases_clean["customer_query"].isna()
)

print(
    "Missing customer queries:",
    service_cases_clean["customer_query_missing_flag"].sum()
)

Missing customer queries: 30


In [165]:
service_cases_clean["customer_query"] = (
    service_cases_clean["customer_query"]
    .fillna("Not Provided")
)

In [166]:
# Negative resolution time flag

service_cases_clean["negative_resolution_flag"] = (
    service_cases_clean["resolution_time_hours"] < 0
)

print(
    "Negative resolution time flags:",
    service_cases_clean["negative_resolution_flag"].sum()
)

Negative resolution time flags: 12


In [167]:
negative_resolution[
    [
        "case_id",
        "customer_id",
        "status",
        "resolution_time_hours",
        "sla_breached"
    ]
]

,case_id,customer_id,status,resolution_time_hours,sla_breached
544,CASE000545,CUST01183,In Progress,-5.0,False
639,CASE000640,CUST00999,In Progress,-8.0,True
1099,CASE001100,CUST00437,Escalated,-6.0,False
1159,CASE001160,CUST00903,Resolved,-5.0,False
1161,CASE001162,CUST00731,Open,-9.0,False
1729,CASE001730,CUST00928,In Progress,-8.0,False
1752,CASE001753,CUST00244,Resolved,-6.0,False
1792,CASE001793,CUST00119,Open,-5.0,False
1823,CASE001824,CUST00472,In Progress,-8.0,True
2100,CASE002101,CUST00237,Open,-1.0,True


In [168]:
# 14.5.6 Convert invalid negative resolution times to missing

service_cases_clean.loc[
    service_cases_clean["negative_resolution_flag"],
    "resolution_time_hours"
] = pd.NA

print(
    "Negative resolution values remaining:",
    (service_cases_clean["resolution_time_hours"] < 0).sum()
)

print(
    "Resolution time missing after correction:",
    service_cases_clean["resolution_time_hours"].isna().sum()
)

Negative resolution values remaining: 0
Resolution time missing after correction: 62


In [169]:
# 14.5.7 Check resolution-time availability by status

resolution_status_check = (
    service_cases_clean
    .groupby("status")["resolution_time_hours"]
    .agg(
        total_cases="count",
        missing_resolution="size"
    )
)

In [170]:
resolution_status_check = (
    service_cases_clean
    .groupby("status")
    .agg(
        total_cases=("case_id", "count"),
        missing_resolution=("resolution_time_hours", lambda x: x.isna().sum())
    )
)

print(resolution_status_check)

             total_cases  missing_resolution
status                                      
Escalated            180                   8
In Progress          611                  23
Open                 418                  10
Resolved            1291                  21


In [171]:
# Flag cases where resolution time is unavailable

service_cases_clean["resolution_time_missing_flag"] = (
    service_cases_clean["resolution_time_hours"].isna()
)

print(
    "Missing/invalid resolution time flags:",
    service_cases_clean["resolution_time_missing_flag"].sum()
)

Missing/invalid resolution time flags: 62


In [172]:
# 14.5.9 Service Cases cleaning summary

print("========== SERVICE CASES CLEANING SUMMARY ==========")

print("Total rows:", len(service_cases_clean))
print("Duplicate case IDs:",
      service_cases_clean["case_id"].duplicated().sum())

print("\nMissing values:")
print(service_cases_clean.isnull().sum())

print("\nQuality flags:")
print("Missing priority:",
      service_cases_clean["priority_missing_flag"].sum())

print("Missing customer query:",
      service_cases_clean["customer_query_missing_flag"].sum())

print("Negative resolution values:",
      service_cases_clean["negative_resolution_flag"].sum())

print("Missing/invalid resolution time:",
      service_cases_clean["resolution_time_missing_flag"].sum())

print("\nStatus values:")
print(service_cases_clean["status"].unique())

print("\nPriority values:")
print(service_cases_clean["priority"].unique())

print("\nData types:")
print(service_cases_clean.dtypes)

print("\n========== CHECK COMPLETED ==========")

========== SERVICE CASES CLEANING SUMMARY ==========
Total rows: 2500
Duplicate case IDs: 0

Missing values:
case_id                          0
customer_id                      0
transaction_id                   0
created_at                       0
case_type                        0
priority                         0
status                           0
resolution_time_hours           62
sla_breached                     0
channel                          0
customer_query                   0
priority_missing_flag            0
customer_query_missing_flag      0
negative_resolution_flag         0
resolution_time_missing_flag     0
dtype: int64

Quality flags:
Missing priority: 37
Missing customer query: 30
Negative resolution values: 12
Missing/invalid resolution time: 62

Status values:
<StringArray>
['In Progress', 'Resolved', 'Open', 'Escalated']
Length: 4, dtype: string

Priority values:
['Low' 'Medium' 'High' 'Unknown' 'Critical']

Data types:
case_id                                 ob

In [173]:
# 14.6 Final Service Cases Quality Check

print("========== FINAL SERVICE CASES QUALITY CHECK ==========")

print("Total rows:", len(service_cases_clean))
print("Total columns:", len(service_cases_clean.columns))

print("\nDuplicate case IDs:",
      service_cases_clean["case_id"].duplicated().sum())

print("\nMissing values:")
print(service_cases_clean.isnull().sum())

print("\nStatus values:")
print(service_cases_clean["status"].unique())

print("\nPriority values:")
print(service_cases_clean["priority"].unique())

print("\nNegative resolution values:",
      (service_cases_clean["resolution_time_hours"] < 0).sum())

print("\nResolution-time missing flags:",
      service_cases_clean["resolution_time_missing_flag"].sum())

print("\nData types:")
print(service_cases_clean.dtypes)

print("\n========== QUALITY CHECK COMPLETED ==========")

========== FINAL SERVICE CASES QUALITY CHECK ==========
Total rows: 2500
Total columns: 15

Duplicate case IDs: 0

Missing values:
case_id                          0
customer_id                      0
transaction_id                   0
created_at                       0
case_type                        0
priority                         0
status                           0
resolution_time_hours           62
sla_breached                     0
channel                          0
customer_query                   0
priority_missing_flag            0
customer_query_missing_flag      0
negative_resolution_flag         0
resolution_time_missing_flag     0
dtype: int64

Status values:
<StringArray>
['In Progress', 'Resolved', 'Open', 'Escalated']
Length: 4, dtype: string

Priority values:
['Low' 'Medium' 'High' 'Unknown' 'Critical']

Negative resolution values: 0

Resolution-time missing flags: 62

Data types:
case_id                                 object
customer_id                           

In [174]:
# 14.7 Export Clean Service Cases Dataset

service_cases_clean.to_csv(
    "service_cases_cleaned.csv",
    index=False
)

print("Clean service cases dataset saved successfully.")
print("File: service_cases_cleaned.csv")
print("Rows:", len(service_cases_clean))
print("Columns:", len(service_cases_clean.columns))

Clean service cases dataset saved successfully.
File: service_cases_cleaned.csv
Rows: 2500
Columns: 15


In [175]:
# Verify exported dataset

service_cases_final = pd.read_csv("service_cases_cleaned.csv")

print("Reloaded rows:", len(service_cases_final))
print("Reloaded columns:", len(service_cases_final.columns))

print(
    "Duplicate case IDs:",
    service_cases_final["case_id"].duplicated().sum()
)

Reloaded rows: 2500
Reloaded columns: 15
Duplicate case IDs: 0


In [176]:
customers_clean.to_csv("customers_cleaned.csv", index=False)

In [177]:
transactions_clean.to_csv("transactions_cleaned.csv", index=False)

In [178]:
fraud_alerts_clean.to_csv("fraud_alerts_cleaned.csv", index=False)

In [179]:
service_cases_clean.to_csv("service_cases_cleaned.csv", index=False)

In [180]:
import os

files = [
    "customers_cleaned.csv",
    "transactions_cleaned.csv",
    "fraud_alerts_cleaned.csv",
    "service_cases_cleaned.csv"
]

for file in files:
    print(file, "->", os.path.exists(file))

customers_cleaned.csv -> True
transactions_cleaned.csv -> True
fraud_alerts_cleaned.csv -> True
service_cases_cleaned.csv -> True


In [181]:
# Export cleaned datasets as real CSV files

customers_clean.to_csv("customers_cleaned.csv", index=False)
transactions_clean.to_csv("transactions_cleaned.csv", index=False)
fraud_alerts_clean.to_csv("fraud_alerts_cleaned.csv", index=False)
service_cases_clean.to_csv("service_cases_cleaned.csv", index=False)

print("All 4 datasets exported successfully as CSV.")

All 4 datasets exported successfully as CSV.


In [182]:
import os

files = [
    "customers_cleaned.csv",
    "transactions_cleaned.csv",
    "fraud_alerts_cleaned.csv",
    "service_cases_cleaned.csv"
]

for file in files:
    print(file, "->", os.path.exists(file))

customers_cleaned.csv -> True
transactions_cleaned.csv -> True
fraud_alerts_cleaned.csv -> True
service_cases_cleaned.csv -> True


In [183]:
# Create NEW CSV files specifically for MySQL import

customers_clean.to_csv("MYSQL_customers.csv", index=False)
transactions_clean.to_csv("MYSQL_transactions.csv", index=False)
fraud_alerts_clean.to_csv("MYSQL_fraud_alerts.csv", index=False)
service_cases_clean.to_csv("MYSQL_service_cases.csv", index=False)

print("CSV files created successfully.")

CSV files created successfully.


In [184]:
import os

for file in [
    "MYSQL_customers.csv",
    "MYSQL_transactions.csv",
    "MYSQL_fraud_alerts.csv",
    "MYSQL_service_cases.csv"
]:
    print(file, "exists:", os.path.exists(file))

MYSQL_customers.csv exists: True
MYSQL_transactions.csv exists: True
MYSQL_fraud_alerts.csv exists: True
MYSQL_service_cases.csv exists: True


In [185]:
import pandas as pd

test = pd.read_csv("MYSQL_customers.csv")

print("Rows:", len(test))
print("Columns:", len(test.columns))
print("File loaded successfully as CSV")

Rows: 1200
Columns: 12
File loaded successfully as CSV


In [186]:
import zipfile
import os

csv_files = [
    "MYSQL_customers.csv",
    "MYSQL_transactions.csv",
    "MYSQL_fraud_alerts.csv",
    "MYSQL_service_cases.csv"
]

with zipfile.ZipFile("MYSQL_CSV_FILES.zip", "w", zipfile.ZIP_DEFLATED) as zipf:
    for file in csv_files:
        if os.path.exists(file):
            zipf.write(file)

print("ZIP file created successfully.")

ZIP file created successfully.
